server/coordinator/procedure/ddl/createpartitiontable/create_partition_table.go

```diff
@@ -219,35 +219,62 @@ func createDataTablesCallback(event *fsm.Event) {
 	}
 
 	shardVersions := req.p.relatedVersionInfo.ShardWithVersion
+	shardTableMetaDatas := make(map[storage.ShardID][]metadata.CreateTableMetadataRequest, 0)
 	for i, subTableShard := range params.SubTablesShards {
-		result, err := params.ClusterMetadata.CreateTableMetadata(req.ctx, metadata.CreateTableMetadataRequest{
+		tableMetaData := metadata.CreateTableMetadataRequest{
 			SchemaName:    params.SourceReq.GetSchemaName(),
 			TableName:     params.SourceReq.GetPartitionTableInfo().SubTableNames[i],
 			PartitionInfo: storage.PartitionInfo{},
-		})
+		}
+		shardTableMetaDatas[subTableShard.ShardInfo.ID] = append(shardTableMetaDatas[subTableShard.ShardInfo.ID], tableMetaData)
+	}
+	shardVersionsCh := make(map[storage.ShardID]chan uint64, len(shardTableMetaDatas))
+	for shardID, tableMetaDatas := range shardTableMetaDatas {
+		shardVersion := shardVersions[shardID]
+		shardVersionsCh[shardID] = make(chan uint64)
+		go createDataTables(event, shardID, tableMetaDatas, shardVersion, shardVersionsCh[shardID])
+	}
+	for shardID, shardVersionCh := range shardVersionsCh {
+		shardVersions[shardID] = <-shardVersionCh
+	}
+}
+
+func createDataTables(event *fsm.Event, shardID storage.ShardID, tableMetaDatas []metadata.CreateTableMetadataRequest, shardVersion uint64, shardVersionCh chan uint64) {
+	defer func() {
+		shardVersionCh <- shardVersion
+	}()
+
+	req, err := procedure.GetRequestFromEvent[*callbackRequest](event)
+	if err != nil {
+		procedure.CancelEventWithLog(event, err, "get request from event")
+		return
+	}
+	params := req.p.params
+
+	for _, tableMetaData := range tableMetaDatas {
+		result, err := params.ClusterMetadata.CreateTableMetadata(req.ctx, tableMetaData)
 		if err != nil {
 			procedure.CancelEventWithLog(event, err, "create table metadata")
 			return
 		}
 
 		shardVersionUpdate := metadata.ShardVersionUpdate{
-			ShardID:     subTableShard.ShardInfo.ID,
-			CurrVersion: req.p.relatedVersionInfo.ShardWithVersion[subTableShard.ShardInfo.ID] + 1,
-			PrevVersion: req.p.relatedVersionInfo.ShardWithVersion[subTableShard.ShardInfo.ID],
+			ShardID:     shardID,
+			CurrVersion: shardVersion + 1,
+			PrevVersion: shardVersion,
 		}
 
-		if err := ddl.CreateTableOnShard(req.ctx, params.ClusterMetadata, params.Dispatch, subTableShard.ShardInfo.ID, ddl.BuildCreateTableRequest(result.Table, shardVersionUpdate, params.SourceReq)); err != nil {
+		if err := ddl.CreateTableOnShard(req.ctx, params.ClusterMetadata, params.Dispatch, shardID, ddl.BuildCreateTableRequest(result.Table, shardVersionUpdate, params.SourceReq)); err != nil {
```

Comment:
We need to support the user to rebuild the table when the table creation fails. You can refer to this pr https://github.com/CeresDB/ceresmeta/pull/179, which supports repeated table creation for ordinary tables.

Reply:
I will solve this problem in the future pr

Reply:
Add a todo comment here.